Arch arm cortex m text section fix

ioannisg opened the pull request:
Rename TEXT_SECTION_NAME_2 to text (TEXT_SECTION_NAME)

Addresses #21650 for Cortex-M

Additionally, fix some wrong pre-processor directives' indentation. [trivial]

ioannisg committed:
arch: arm: cortex-m: linker.ld: do not indent preprocessor directives

We should not use indentation for pre-processor directives.
This commit fixes the indentation in the ARM Cortex-M linker
script.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: cortex-m: properly name ROM sections in the linker

This commit does the following:
- renames the 'text' ROM section to 'rom_start', to reflect
  that this section is the first section of the image.
- renames the 'TEXT_SECTION_NAME_2' section to 'text', since
  that section (whose start is pointed by _image_text_start)
  holds the entire image text section.

The commit removes the confusion by having multiple ROM sections
named as 'text' in ARM Cortex-M builds.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

andrewboie commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -40,30 +40,30 @@
-  #define ROM_SIZE (CONFIG_FLASH_SIZE*1K - CONFIG_FLASH_LOAD_OFFSET - \
-		    CCFG_SIZE)
-  #define CCFG_ADDR (ROM_ADDR + ROM_SIZE)
+#define CCFG_SIZE 88
```
> We should not use indentation for pre-processor directives.

Why not? IMO this reduces readability when we have a bunch of nested clauses

ioannisg replied:
This issue came up in another review; apparently nested pre-processor directives are not indented for ~90% of the code in Zephyr, so I wanted to do a step forward into this matter, and enforce consistency in the ARM linker script. If we agree to indent all nested #if and #define directives, we should have a rule for that....

tyrantlizard77 reviewed (dismissed):
Not thrilled with the indentation changed, but not enough to -1.

Other patch LGTM, assuming sanitycheck doesn't throw errors about extra sections and this doesn't mess with MPU configuration.

andrewboie commented:
Could we just reverse the order of the text and data sections? and coalesce the vector table and ARM exception unwinding sections into rodata?

ioannisg commented:
> Could we just reverse the order of the text and data sections? and coalesce the vector table and ARM exception unwinding sections into rodata?

In principle this could be done @andrewboie ; I am just wondering why we would like to do this for ARM Cortex-M; apparently there's a similar sructure in other ARCHEs; text section comes before rodata....

andrewboie commented:
> apparently there's a similar sructure in other ARCHEs; text section comes before rodata...

There's no requirement to what order they appear in. I reordered some data sections on 32-bit x86 to solve a technical problem.

ioannisg commented:
> > apparently there's a similar sructure in other ARCHEs; text section comes before rodata...
> 
> There's no requirement to what order they appear in. I reordered some data sections on 32-bit x86 to solve a technical problem.

I understand. But what is the problem **here**, with the current ordering of the sections? If reordering does an improvement, I am up for it. :)

andrewboie commented:
I'm not saying there's a *problem*, but I think reversing the sections is *simpler*

ioannisg commented:
@wentongwu  could you please review this patch (the one that renames the .text section) and see if this will work with Code Relocation?

andrewboie commented:
@wentongwu can you take a look at this?

switching the rodata and text areas had side effects, this approach is OK with me

oyvindronningstad commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -135,7 +135,8 @@ SECTIONS
 
-    SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
+    /* The exception vector table is located at the beginning of ROM */
+    SECTION_PROLOGUE(vector_table,,)
```
riscv has the same scheme, but calls the section "vector", maybe align on that name?

ioannisg replied:
I think the name is good; it is actually the ARM Cortex-M exception vector table; maybe we could rename the Risc-V Section instead, but that's another story.

oyvindronningstad reviewed (commented):
The `#include <snippets-text-start.ld>` is now no longer in the text section, so you should move it. Unfortunately, this means moving the contents of aarch32/vector_table.ld, text_section_offset.ld (keep this file still for riscv and x86), and cc32xx_debug.ld back into the new vector table section.

ioannisg committed:
arch: arm: cortex-m: move openocd sections after _vector_end

openocd linker sections are not supposed to be part of the
vector table sections. Place the sections after we define
the _vector_end linker symbol.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

oyvindronningstad committed:
extensions.cmake: Replace TEXT_START with ROM_START

In zephyr_linker_sources().
This is done since the point of the location is to place things at given
offsets. This can only be done consistenly if the linker code is placed
into the _first_ section.

All uses of TEXT_START are replaced with ROM_START.

ROM_START is only supported in some arches, as some arches have several
custom sections before text. These don't currently have ROM_START or
TEXT_START available, but that could be added with a bit of refactoring
in their linker script.

No SORT_KEYs are changed.

This also fixes an error introduced when TEXT_START was added, where
TEXT_SECTION_OFFSET was applied to riscv's common linker.ld instead of
to openisa_rv32m1's specific linker.ld.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad commented:
I'm working on this PR.

ioannisg commented on arch/common/text_section_offset_section.ld:
```diff
@@ -0,0 +1,11 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+SECTION_PROLOGUE(text_section_offset,,)
```
There is text_section_offset.ld file in arch/common, do we need to have both @oyvindronningstad ?

oyvindronningstad replied:
Yes, one is for ARM, the other is for RISCV and x86.

ioannisg replied:
OK, saw it. I think we need to implement this a bit better.

oyvindronningstad replied:
Do you want more besides the comments you requested in the other comment?

ioannisg commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -135,26 +135,20 @@ SECTIONS
+/* Located in generated directory. This file is populated by the
+ * zephyr_linker_sources() Cmake function.
+ */
+#include <snippets-rom-start.ld>
```
@oyvindronningstad it is now essential to add some information here about the sections before TEXT: what do they, generally contain? We need to say they contain the vector table, at least, so people won't be wondering where is the default Cortex-M exception table

oyvindronningstad replied:
I added something to the comment. Note that I added a commit changing the wording of every comment to make them mention the relevant argument to `zephyr_linker_sources()`

ioannisg commented on arch/common/CMakeLists.txt:
```diff
@@ -34,7 +34,9 @@ zephyr_linker_sources_ifdef(CONFIG_NOCACHE_MEMORY
+if (DEFINED CONFIG_X86 OR DEFINED CONFIG_RISCV)
   zephyr_linker_sources(TEXT_START SORT_KEY 0x0 text_section_offset.ld)
+elseif (DEFINED CONFIG_ARM)
+  zephyr_linker_sources(ROM_START  SORT_KEY 0x0 text_section_offset_section.ld)
```
This thing needs to be re-factored properly, IMO.
- we need to comment inline why the ARM text section offset is different from RISC-V, X86
- we need to try to combine have a single file for all these ARCHes to implement the section offset.

oyvindronningstad replied:
Do you want the comments here in the CMakeLists.txt file?

ioannisg replied:
Ideally, I would like _not_ to have the if, else if in CmakeLists, and apply any required conditioning inside the single text_Section_offset.ld

ioannisg replied:
update: you could keep the if, else here, but is it possible to have the same file linked, i.e. text_Section_offset.ld?

oyvindronningstad replied:
Merged into one ifdefed file

vanti reviewed (changes requested):
I tried this on cc3220sf_launchxl, and got a linker error:

```
$ west build -b cc3220sf_launchxl
-- west build: build configuration:
       source directory: /home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get
       build directory: /home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get/build
       BOARD: cc3220sf_launchxl (origin: CMakeCache.txt)
-- west build: building application
[1/6] Linking C executable zephyr/zephyr_prebuilt.elf
FAILED: zephyr/zephyr_prebuilt.elf 
: && ccache /home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -Wl,-T zephyr/linker.cmd -Wl,-Map=/home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get/build/zephyr/zephyr_prebuilt.map -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/common/libarch__common.a zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a zephyr/lib/libc/newlib/liblib__libc__newlib.a zephyr/lib/posix/liblib__posix.a zephyr/boards/arm/cc3220sf_launchxl/libboards__arm__cc3220sf_launchxl.a zephyr/subsys/net/libsubsys__net.a zephyr/subsys/net/l2/wifi/libsubsys__net__l2__wifi.a zephyr/subsys/net/ip/libsubsys__net__ip.a zephyr/drivers/gpio/libdrivers__gpio.a zephyr/drivers/serial/libdrivers__serial.a modules/ti/simplelink/source/ti/devices/cc32xx/lib..__modules__hal__ti__simplelink__source__ti__devices__cc32xx.a modules/ti/simplelink/lib..__modules__hal__ti__simplelink.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -L"/home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/thumb/v7e-m/nofp" -L/home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get/build/zephyr -lgcc -Wl,--print-memory-usage -mthumb -mcpu=cortex-m4 -Wl,--gc-sections -Wl,--build-id=none -Wl,--sort-common=descending -Wl,--sort-section=alignment -Wl,-u,_OffsetAbsSyms -Wl,-u,_ConfigAbsSyms -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--orphan-handling=warn -mabi=aapcs -march=armv7e-m -lm -lc -lgcc -L"/home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/arm-zephyr-eabi"/lib/thumb/v7e-m/nofp && :
/home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld:zephyr/linker.cmd:45: syntax error
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

I have attached linker.cmd before and after the change.
[linker.cmd.after.txt](https://github.com/zephyrproject-rtos/zephyr/files/4087618/linker.cmd.after.txt)
[linker.cmd.before.txt](https://github.com/zephyrproject-rtos/zephyr/files/4087619/linker.cmd.before.txt)

oyvindronningstad commented:
> I tried this on cc3220sf_launchxl, and got a linker error:
> 
> ```
> $ west build -b cc3220sf_launchxl
> -- west build: build configuration:
>        source directory: /home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get
>        build directory: /home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get/build
>        BOARD: cc3220sf_launchxl (origin: CMakeCache.txt)
> -- west build: building application
> [1/6] Linking C executable zephyr/zephyr_prebuilt.elf
> FAILED: zephyr/zephyr_prebuilt.elf 
> : && ccache /home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr/zephyr_prebuilt.elf  -Wl,-T zephyr/linker.cmd -Wl,-Map=/home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get/build/zephyr/zephyr_prebuilt.map -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/common/libarch__common.a zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a zephyr/lib/libc/newlib/liblib__libc__newlib.a zephyr/lib/posix/liblib__posix.a zephyr/boards/arm/cc3220sf_launchxl/libboards__arm__cc3220sf_launchxl.a zephyr/subsys/net/libsubsys__net.a zephyr/subsys/net/l2/wifi/libsubsys__net__l2__wifi.a zephyr/subsys/net/ip/libsubsys__net__ip.a zephyr/drivers/gpio/libdrivers__gpio.a zephyr/drivers/serial/libdrivers__serial.a modules/ti/simplelink/source/ti/devices/cc32xx/lib..__modules__hal__ti__simplelink__source__ti__devices__cc32xx.a modules/ti/simplelink/lib..__modules__hal__ti__simplelink.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -L"/home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/thumb/v7e-m/nofp" -L/home/local/zephyrproject-work/zephyr/samples/net/sockets/http_get/build/zephyr -lgcc -Wl,--print-memory-usage -mthumb -mcpu=cortex-m4 -Wl,--gc-sections -Wl,--build-id=none -Wl,--sort-common=descending -Wl,--sort-section=alignment -Wl,-u,_OffsetAbsSyms -Wl,-u,_ConfigAbsSyms -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--orphan-handling=warn -mabi=aapcs -march=armv7e-m -lm -lc -lgcc -L"/home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/arm-zephyr-eabi"/lib/thumb/v7e-m/nofp && :
> /home/local/bin/zephyr-sdk-0.10.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/../../../../arm-zephyr-eabi/bin/ld:zephyr/linker.cmd:45: syntax error
> collect2: error: ld returned 1 exit status
> ninja: build stopped: subcommand failed.
> ```
> 
> I have attached linker.cmd before and after the change.
> [linker.cmd.after.txt](https://github.com/zephyrproject-rtos/zephyr/files/4087618/linker.cmd.after.txt)
> [linker.cmd.before.txt](https://github.com/zephyrproject-rtos/zephyr/files/4087619/linker.cmd.before.txt)

Thanks, @vanti. I fixed it and tested locally, it should be working now.

vanti commented:
> Thanks, @vanti. I fixed it and tested locally, it compiles cleanly now.

Thanks, it does compile cleanly now. When I tried to run the executable however, it does not run correctly. From the code placement in the map file, I see that the cc32xx_debug section is now occupying 0xc, moving everything else forward:

```
cc32xx_debug    0x0000000001000000        0xc
                0x0000000000000000                . = 0x0
 *(SORT_BY_ALIGNMENT(.dbghdr))
 .dbghdr        0x0000000001000000        0xc zephyr/boards/arm/cc3220sf_launchxl/libboards__arm__cc3220sf_launchxl.a(dbghdr.c.obj)
                0x0000000001000000                ulDebugHeader
 *(SORT_BY_ALIGNMENT(.dbghdr.*))

text_section_offset
                0x000000000100000c      0x800
                0x0000000000000800                . = 0x800
 *fill*         0x000000000100000c      0x800 
                0x000000000100080c                . = ALIGN (0x4)

vector_table    0x000000000100080c      0x308
                0x000000000100080c                _vector_start = .
 *(SORT_BY_ALIGNMENT(.exc_vector_table))
 *(SORT_BY_ALIGNMENT(.exc_vector_table.*))
 .exc_vector_table._vector_table_section
```

Here is how it was before the change:

```
text            0x0000000001000000      0xb08
                0x0000000000000000                . = 0x0
 *(SORT_BY_ALIGNMENT(.dbghdr))
 .dbghdr        0x0000000001000000        0xc zephyr/boards/arm/cc3220sf_launchxl/libboards__arm__cc3220sf_launchxl.a(dbghdr.c.obj)
                0x0000000001000000                ulDebugHeader
 *(SORT_BY_ALIGNMENT(.dbghdr.*))
                0x0000000000000800                . = 0x800
 *fill*         0x000000000100000c      0x7f4 
                0x0000000001000800                . = ALIGN (0x4)
                0x0000000001000800                _vector_start = .
 *(SORT_BY_ALIGNMENT(.exc_vector_table))
 *(SORT_BY_ALIGNMENT(.exc_vector_table.*))
 .exc_vector_table._vector_table_section
```

Is it possible to preserve the original placement?

oyvindronningstad commented:
> Is it possible to preserve the original placement?

Thanks again @vanti and sorry I missed this.

This exposes a bigger problem, namely that moving the location counter (".") to a given offset from the start of the image is broken with this. It's possible to fix by changing

    . = CONFIG_TEXT_SECTION_OFFSET;

to

    . = ABSOLUTE(CONFIG_TEXT_SECTION_OFFSET + _image_rom_start);

but this I feel is quite an invasive change, since every place that moves the location counter has to be changed in the same way.

I suggest we go back to Ioannis' proposal, but call the section "rom_start", so it is defined as the "first" section, which makes it safe to put input sections there if they need such an offset.

oyvindronningstad commented:
This is now reverted back to Ioannis's proposal, with an added commit to rename TEXT_START to ROM_START and remove it from some arches (since they have very specific sections as their first section).

ioannisg commented:
> This is now reverted back to something close to Ioannis's initial proposal, with an added commit to rename TEXT_START to ROM_START and remove it from some arches (since they have very specific sections as their first section).

@SebastianBoe @hakonfam could you revisit this, since it has been re-done?

vanti reviewed (approved):
Thanks for the fix. It now works on SimpleLink. LGTM!

hakonfam commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -151,7 +151,7 @@ SECTIONS
 #endif /* CONFIG_CODE_DATA_RELOCATION */
 
-    SECTION_PROLOGUE(_TEXT_SECTION_NAME_2,,)
+    SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
```
Should we do the same fix for cortex-r ?

ioannisg replied:
Will do that on separate PR, hopefully by the cortex r contributors

hakonfam commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -151,7 +151,7 @@ SECTIONS
 #endif /* CONFIG_CODE_DATA_RELOCATION */
 
-    SECTION_PROLOGUE(_TEXT_SECTION_NAME_2,,)
+    SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
```
I see that `_TEXT_SECTION_NAME_2` is used in `scripts/sanitycheck` for doing some size calculation. Should the script be updated to reflect these changes?

ioannisg replied:
Yeah, we left it there because we don't fix the cortex-r now. Will make sure I ask @stephanosio for the cortex r fix ::)

SebastianBoe replied:
Does this break the size calculation? What consequences does this have?

ioannisg replied:
@SebastianBoe I am not sure where "_this_" exactly refers to; i'll try to reply, anyways.

The current patch only renames the "TEXT_SECTION_NAME_2" section to "text", and the original "text" section to rom_start; note that _that_ section contains the vector table.

The patch does not move any symbols between sections. 

Where does size calculation take place?

SebastianBoe replied:
By this I meant this patch. Will see if I can find the relevant sanitycheck code.

SebastianBoe replied:
It appears that sanitycheck believes that TEXT_SECTION_NAME_2 is in the rw area.

Which I assume is wrong, so this patch should not cause any regression, it might even fix
the calculation.

https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/sanitycheck#L1074
